### ISSUE TITLE
Improve handling of network outages during transfer

### ISSUE THREAD

**Issue by tomck** [kblin/ncbi-genome-download#12]
While downloading all bacteria, I received the above error, both when running it the first time and when running it the second (correct the mismatches) time. Not sure if there's an actual problem or not.

```
bash-3.2$ ncbi-genome-download bacteria --parallel 5
ERROR: No entry for file ending in '_genomic.gbff.gz'
ERROR: Checksum mismatch for u'$HOME/tmp/bacteria/refseq/bacteria/GCF_000986765.1/GCF_000986765.1_ASM98676v1_genomic.gbff.gz'. Expected u'06e551ac87e510f4275fcd04302277b7', got 'ed004afb30643a1cae37e01d7fa93523'
ERROR: Checksum mismatch for u'$HOME/tmp/bacteria/refseq/bacteria/GCF_001650315.1/GCF_001650315.1_ASM165031v1_genomic.gbff.gz'. Expected u'fd17ff629c497bd8de1715f35a1c04eb', got '691d3769cb390b10495c5235b40bb993'
ERROR: Checksum mismatch for u'$HOME/tmp/bacteria/refseq/bacteria/GCF_000633455.1/GCF_000633455.1_de_novo_genomic.gbff.gz'. Expected u'b7b1441a00d8983e7ad0bcebc3c031e1', got '1e4b8ce848e5fd22caaed4ba76bd1f9f'
bash-3.2$ ncbi-genome-download bacteria --parallel 5
ERROR: No entry for file ending in '_genomic.gbff.gz'
bash-3.2$ python --version
Python 2.7.11
```

**Comment by kblin** [comment]
Hi,

could you re-run with `--debug` and post the log here?

**Comment by tomck** [comment]
I snipped most of the log but have the [entire thing](http://pastebin.com/1rre8wtJ) (did 2> ncbidebug.txt) if you need it. 

```
INFO: Downloading record u'GCF_001658255.1'
INFO: Starting new HTTP connection (1): ftp.ncbi.nlm.nih.gov
DEBUG: "GET /genomes/all/GCF_001658255.1_ASM165825v1/md5checksums.txt HTTP/1.1" 200 1794
DEBUG: Skipping over unexpected checksum line u''
ERROR: No entry for file ending in '_genomic.gbff.gz'
INFO: Downloading record u'GCF_001658265.1'
INFO: Starting new HTTP connection (1): ftp.ncbi.nlm.nih.gov
Traceback (most recent call last):
  File "/usr/local/bin/ncbi-genome-download", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/site-packages/ncbi_genome_download/__main__.py", line 59, in main
    ncbi_genome_download.download(args)
  File "/usr/local/lib/python2.7/site-packages/ncbi_genome_download/core.py", line 50, in download
    args.assembly_level, args.genus, args.parallel)
  File "/usr/local/lib/python2.7/site-packages/ncbi_genome_download/core.py", line 67, in _download
    download_jobs.extend(download_entry(entry, section, domain, output, file_format))
  File "/usr/local/lib/python2.7/site-packages/ncbi_genome_download/core.py", line 98, in download_entry
    checksums = grab_checksums_file(entry)
  File "/usr/local/lib/python2.7/site-packages/ncbi_genome_download/core.py", line 140, in grab_checksums_file
    req = requests.get(full_url)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 71, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/api.py", line 57, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/sessions.py", line 585, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/requests/adapters.py", line 467, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='ftp.ncbi.nlm.nih.gov', port=80): Max retries exceeded with url: /genomes/all/GCF_001658265.1_ASM165826v1/md5checksums.txt (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x10f3e6e10>: Failed to establish a new connection: [Errno 65] No route to host',))
```

**Comment by kblin** [comment]
Can't reproduce here, GCF_001658255.1_ASM165825v1_genomic.gbff.gz exists and downloads fine.
However, you seem to be having networking problems with that "no route to host".

**Comment by tomck** [comment]
I realize that, my assumption is the program doesn't handle temporary network outages very well, because I have to run from the beginning when trying again and no guarantees it will finish.

**Comment by kblin** [comment]
I'm not sure what you expect the program to do. I certainly can fail immediately if there's any network connectivity issue, but you'd still need to be running the download tool until it returns successfully. Would that help?

**Comment by tomck** [comment]
For my specific case, I run the app in screen, so if it just kept retrying with perhaps a slight backoff that would be fine. Long term, maybe some sort of --resume option with a status file?

**Comment by kblin** [comment]
Apart from saving you from reading the `assembly_summary.txt` file and a couple of MD5SUMS files again, this doesn't drastically change what ncbi-genome-download will do. If the md5sum of your local file matches the md5sum in the file from NCBI, no new download will be started.
At no point will ncbi-genome-download re-download a file when the checksum matches the NCBI files.
So --resume would basically achieve the same thing, wouldn't it? Apart from checking that your other files are still correct, that is.

**Comment by tomck** [comment]
I don't seem to have an assembly_summary.txt file, but yes, --resume does sound unnecessary. However the retry-until-finished doesn't sound too bad, does it?

**Comment by kblin** [comment]
The assembly_summary.txt file isn't stored locally, it's just parsed from NCBI.

I'll see what I can do on the retry front, but unfortunately that's gotten a bit more tricky with the parallel download support.

**Comment by kblin** [comment]
Should be solved by a3f7cb1, but it's a bit difficult to test. Please reopen if this doesn't address your problem.
